File: my_code/DSA/JS/DS/linked-lists/doubly-linked-list/circular-dobuly-linked-list.js
class Node {
  constructor(val) {
    this.val = val;
    this.next = null;
    this.prev = null;
  }
}

class DoublyLinkedList {
  constructor() {
    this.head = null;
    this.tail = null;
    this.length = 0;
  }

  // add a new node at the end of the dll
  push(val) {
    // make the new node
    const newNode = new Node(val);

    if (!this.head) {
      // if empty => make the head and the tail to be the newNode
      this.head = newNode;
      this.tail = newNode;
      // circular links
      this.head.prev = this.tail;
      this.tail.next = this.head;
    } else {
      // make the tail point into the new node
      this.tail.next = newNode;
      // set the new node to point back to the tail
      newNode.prev = this.tail;
      // update the tail to be the new node
      this.tail = newNode;
      // maintain circular links
      this.tail.next = this.head;
      this.head.prev = this.tail;
    }
    // always increment the length
    this.length++;
    // return the new list
    return this;
  }

  //remove the last node
  pop() {
    // if empty
    if (!this.head) {
      return undefined;
    }

    // store it to return it after removing
    const poppedNode = this.tail;

    // if just one node => return to the initial point
    if (this.length === 1) {
      this.head = null;
      this.tail = null;
    }
    //
    else {
      // set the tail to be the previous node
      this.tail = poppedNode.prev;
      // set the new tail next to point into head (circular)
      this.tail.next = this.head;
      // set the head prev to the new tail (circular)
      this.head.prev = this.tail;
      // unlink the popped node
      poppedNode.prev = null;
      poppedNode.next = null;
    }
    // decrement the length
    this.length--;
    // return the popped node
    return poppedNode;
  }

  // remove the head
  shift() {
    if (!this.head) {
      return undefined;
    }

    // store the head to return it later
    const oldHead = this.head;

    // if just one node => return to the initial point
    if (this.length === 1) {
      this.head = null;
      this.tail = null;
    } else {
      // set the head to be the new one (second node)
      this.head = oldHead.next;
      // set the new head prev to point to tail (circular)
      this.head.prev = this.tail;
      // set the tail next to new head (circular)
      this.tail.next = this.head;
      // unlink the old head
      oldHead.next = null;
      oldHead.prev = null;
    }
    this.length--;
    return oldHead;
  }

  // add a new node as head
  unshift(val) {
    // first create a new node to add it
    const newNode = new Node(val);

    // if empty make the head and tail to be the new node
    if (!this.head) {
      this.head = newNode;
      this.tail = newNode;
      // circular links
      this.head.prev = this.tail;
      this.tail.next = this.head;
    } else {
      // set the new node next to point into the old head
      newNode.next = this.head;
      // then set the old head prev to point back into the new node
      this.head.prev = newNode;
      // then update the head to be the new node
      this.head = newNode;
      // maintain circular links
      this.head.prev = this.tail;
      this.tail.next = this.head;
    }
    // increment the length
    this.length++;
    // then return the list
    return this;
  }

  // get a node based on specific index
  get(index) {
    // invalid index
    if (index < 0 || index >= this.length) {
      return null;
    }

    let current, counter;
    // if closer to the head
    // start from the head
    if (index <= this.length / 2) {
      current = this.head;
      counter = 0;
      while (counter < index) {
        current = current.next;
        counter++;
      }
      return current;
    }
    // if closer to the tail
    // start from the tail
    else {
      current = this.tail;
      counter = this.length - 1;
      while (counter > index) {
        current = current.prev;
        counter--;
      }
      return current;
    }
  }

  // set a node value with a new one
  set(index, val) {
    // use the get method to get the node
    const foundNode = this.get(index);

    // if the node exist
    if (foundNode) {
      // set its value to be the new one
      foundNode.val = val;
      return true;
    } else {
      return false;
    }
  }

  // insert a new node based on index
  insert(index, val) {
    // invalid index
    if (index < 0 || index > this.length) {
      return false;
    }
    // if the first node => unshift
    else if (index === 0) {
      this.unshift(val);
      return true;
    }
    // if the last node length => push
    else if (index === this.length) {
      this.push(val);
      return true;
    } else {
      // create a new node
      const newNode = new Node(val);

      // get the node before index we want add
      const beforeNode = this.get(index - 1);
      // get the node after (more readable)
      const afterNode = beforeNode.next;

      //  set the new node next to point into the after node
      newNode.next = afterNode;
      //  and the after node prev to point back to the new node
      afterNode.prev = newNode;
      //  set the before node next to point into new node
      beforeNode.next = newNode;
      // and the new node prev to point back into the before node
      newNode.prev = beforeNode;

      this.length++;
      return true;
    }
  }

  // remove a node based on specific index
  remove(index) {
    // invalid index here also the length not valid
    if (index < 0 || index >= this.length) {
      return undefined;
    }
    // remove from first
    else if (index === 0) {
      return this.shift();
    }
    // remove from last
    else if (index === this.length - 1) {
      return this.pop();
    }
    // remove from inside the list
    else {
      // get the node will remove
      let foundNode = this.get(index);
      // the one before it
      let beforeNode = foundNode.prev;
      // the one after it
      let afterNode = foundNode.next;

      // then make the before node jump and point on after node
      beforeNode.next = afterNode;
      // then make the after node jump and point back to before node
      afterNode.prev = beforeNode;

      // unlink the removed node
      foundNode.next = null;
      foundNode.prev = null;

      this.length--;
      return foundNode;
    }
  }

  // print
  print() {
    let res = "";
    if (!this.head) {
      console.log(res);
      return;
    }
    let current = this.head;
    do {
      res += current.val + " ";
      current = current.next;
    } while (current !== this.head);
    console.log(res);
  }
}

let list = new DoublyLinkedList();
list.push(10).push(20).push(30);
list.print();
list.remove(1);
list.print();

File: my_code/DSA/PYTHON/DS/linked lists/doubly-linked-list/circular-dobuly-linked-list.py
class Node:
    def __init__(self, val):
        self.val = val
        self.next = None
        self.prev = None

class DoublyLinkedList:
    def __init__(self):
        self.head = None
        self.tail = None
        self.length = 0

    # add a new node at the end of the dll
    def push(self, val):
        new_node = Node(val)

        if not self.head:
            self.head = new_node
            self.tail = new_node
            # circular link
            self.head.prev = self.tail
            self.tail.next = self.head
        else:
            # set the tail to point into the new_node
            self.tail.next = new_node
            # and set the new_node to point back to the tail
            new_node.prev = self.tail
            # then update the tail
            self.tail = new_node
            # maintain circular links
            self.tail.next = self.head
            self.head.prev = self.tail
        
        self.length += 1
        return self
    
    # remove the last node
    def pop(self):
        # if empty
        if not self.head:
            return None
        
        # store it to return it after removing
        popped_node = self.tail

        # if just one node => return to the initial point
        if self.length == 1:
            self.head = None
            self.tail = None
        else:
            # set the tail to be the previous node
            self.tail = popped_node.prev
            # set the new tail next to point into head (circular)
            self.tail.next = self.head
            # set the head prev to point back to new tail
            self.head.prev = self.tail
            # unlink popped_node
            popped_node.prev = None
            popped_node.next = None

        self.length -= 1
        # return the popped node
        return popped_node

    # remove the head
    def shift(self):
        if not self.head:
            return None
        
        # store the head to return it later
        old_head = self.head

        # if just one node => return to the initial point
        if self.length == 1:
            self.head = None
            self.tail = None
        else:
            # set the head to be the new one (second node)
            self.head = old_head.next

            # set the new head prev to point to tail (circular)
            self.head.prev = self.tail
            # set the tail next to point to new head (circular)
            self.tail.next = self.head

            # set the old head next and prev to None (unlink)
            old_head.next = None 
            old_head.prev = None
        
        self.length -= 1
        return old_head

    # add a new node as head
    def unshift(self, val):
        # first create a new node to add it
        new_node = Node(val)

        # if empty make the head and tail to be the new node
        if not self.head:
            self.head = new_node
            self.tail = new_node
            # circular links
            self.head.prev = self.tail
            self.tail.next = self.head
        else:
            # set the new node next to point into the old head
            new_node.next = self.head
            # then set the old head prev to point into the new node
            self.head.prev = new_node
            # then update the head to be the new node
            self.head = new_node
            # maintain circular links
            self.tail.next = self.head
            self.head.prev = self.tail

        # increment one to the length
        self.length += 1
        # return the list
        return self

    # get a node based on specific index
    def get(self, index):
        # invalid index
        if index < 0 or index >= self.length:
            return None
        
        # if closer to the head
        # start from the head
        if index <= self.length // 2:
            current = self.head
            counter = 0
            while counter != index:
                current = current.next
                counter += 1
            return current
        # if closer to the tail
        # start from the tail
        else:
            current = self.tail
            counter = self.length - 1
            while counter != index:
                current = current.prev
                counter -= 1
            return current

    # set a node with a new value
    def set(self, index, value):
        # use the get method to get the node
        found_node = self.get(index)

        # if the node exist
        if found_node:
            # set its value to be the new one
            found_node.val = value
            return True
        else:
            return False

    # insert a new node based on index
    def insert(self, index, val):
        # invalid index
        if index < 0 or index > self.length:
            return False
        # if the first node => unshift
        elif index == 0:
            self.unshift(val)
            return True
        # if the last node length => push
        elif index == self.length:
            self.push(val)
            return True
        
        else:
            # create a new node
            new_node = Node(val)
            # get the node before index we want add
            before_node = self.get(index - 1)
            # get the node after (more readable)
            after_node = before_node.next

            # set the new node next to point into the after node
            new_node.next = after_node
            # and the after node prev to point back to the new node
            after_node.prev = new_node
            # set the before node next to point into new node
            before_node.next = new_node
            # and the new node prev to point back into the before node
            new_node.prev = before_node

            self.length += 1
            return True


    # remove a node based on the given index
    def remove(self, index):
        # invalid index here also the length not valid
        if index < 0 or index >= self.length:
            return None
        #  remove from first
        elif index == 0:
            return self.shift()
        #  remove from last
        elif index == self.length - 1:
            return self.pop()
        # remove from inside the list
        else:
            # get the node will remove
            removed_node = self.get(index)
            # the one before it
            before_node = removed_node.prev
            #  the one after it
            after_node = removed_node.next

            #  then make the before node jump and point on after node
            before_node.next = after_node
            #  then make the after node jump and point back to before node
            after_node.prev = before_node

            #  unlink the removed node
            removed_node.next = None
            removed_node.prev = None

            self.length -= 1
            return removed_node

    # to check if the linked list tail refer to the head
    # without using tail (harder)
    def is_perfectly_circular(self):
        if not self.head:
            return False
        
        current = self.head.next
        while current and current != self.head:
            current = current.next

        if current == self.head:
            return True
        
    # Traversing
    def __str__(self):
        if not self.head:
            return ''
        res = ''
        current = self.head
        while True:
            res += str(current.val) + ' '
            current = current.next
            if current == self.head:
                break
        return res.strip()

# Example
l = DoublyLinkedList()
l.push(10).push(20).push(30)
print(l)  # Output: 10 20 30 
l.remove(2)
print(l)  # Output: 10 20
print(l.is_perfectly_circular())
